drivers/audio/pdm/pdm_nrfx.c

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2020 Oticon A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <nrfx_pdm.h>
+#include <drivers/audio/pdm.h>
+#include <hal/nrf_gpio.h>
+#include <stdbool.h>
+
+#define LOG_LEVEL CONFIG_PDM_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(pdm_nrfx);
+
+#define DT_DRV_COMPAT nordic_nrf_pdm
+DEVICE_DECLARE(pdm_0);
+
+#define CLK_PIN       DT_PROP(DT_NODELABEL(pdm), clk_pin)
+#define DIN_PIN       DT_PROP(DT_NODELABEL(pdm), din_pin)
+
+#define PDM_STACK_SIZE 256
+#define PDM_PRIORITY   5
+
+#define PDM_NRFX_NUMBER_OF_BUFFERS 2
+#define PDM_NRFX_BUFFER_SIZE       CONFIG_PDM_BUFFER_SIZE
+
+static K_THREAD_STACK_DEFINE(pdm_stack, PDM_STACK_SIZE);
+
+static struct k_work_q m_pdm_work_q;
+
+struct pdm_nrfx_config {
+	/**
+	 * Default configuration:
+	 * mode			= MONO
+	 * edge			= LEFTFALLING
+	 * clock_freq		= 1032K / 1.032 MHz
+	 * gain_l		= 0x28
+	 * gain_r		= 0x28
+	 * interrupt_priority	= 7
+	 */
+	nrfx_pdm_config_t       default_config;
+};
+
+struct buffer_released_data {
+	struct k_work work;
+	int16_t *buffer;
+};
+
+struct pdm_nrfx_data {
+	pdm_data_handler_t data_handler;
+	struct k_mem_slab *mem_slab;
+	struct buffer_released_data buff_released;
+	uint8_t active_buffer;
+};
+
+static int16_t m_next_buffer[PDM_NRFX_BUFFER_SIZE][PDM_NRFX_NUMBER_OF_BUFFERS];
+
+static struct pdm_nrfx_data m_pdm_nrfx_data;
+
+static void release_buffer(struct k_work *item)
+{
+	struct device *dev = DEVICE_GET(pdm_0);
+
+	struct pdm_nrfx_data *driver_data = dev->driver_data;
+
+	struct buffer_released_data *data =
+		CONTAINER_OF(item, struct buffer_released_data, work);
+
+	driver_data->data_handler(data->buffer, PDM_NRFX_BUFFER_SIZE);
+}
+
+static void pdm_nrfx_event_handler(nrfx_pdm_evt_t const *p_evt)
+{
+	struct device *dev = DEVICE_GET(pdm_0);
+	struct pdm_nrfx_data *driver_data = dev->driver_data;
+
+	void *buffer;
+
+	if (p_evt->error == NRFX_PDM_ERROR_OVERFLOW) {
+		LOG_ERR("Overflow error when handling event.\n");
+		return;
+	}
+
+	if (nrfx_pdm_enable_check()) {
```

Comment:
Better to do the reverse check and exit early if `nrfx_pdm_enable_check` returns false. Helps to decrease cascading.

```
if (nrfx_pdm_enable_check() == false)
    return; 
```